### ISSUE TITLE
Entire install is broken after an insufficient storage error during update

### ISSUE THREAD

**Issue by PleaseConverge** [conda/conda#13143]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

While I knew I was running out of space on my main partition, I thought I still had at least a couple of gigs left. 
As it turned out, I didn't.

When I ran `conda update --all` for the first time in a couple of months, there were 175 packages and dependencies to update, which was enough to consume all the space I had left on the partition. Conda threw an insufficient storage error while downloading the packages, as well a couple of warnings along the line of "this package may need to be redownloaded".

After freeing up ~5GB of space, I reran `conda update --all`, it listed the same 175 packages as before, and finished without any error or warning.

As far as I can tell, everything is now broken:

<details><summary>When I try anything with the `conda` command, I get:</summary>
<p>

```
Traceback (most recent call last):
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/exception_handler.py", line 17, in __call__
    return func(*args, **kwargs)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/cli/main.py", line 54, in main_subshell
    parser = generate_parser(add_help=True)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/cli/conda_argparse.py", line 115, in generate_parser
    configure_parser_create(sub_parsers)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/cli/conda_argparse.py", line 767, in configure_parser_create
    add_parser_solver(solver_mode_options)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/cli/conda_argparse.py", line 1975, in add_parser_solver
    solver.name for solver in context.plugin_manager.get_hook_results("solvers")
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/base/context.py", line 492, in plugin_manager
    from ..plugins.manager import get_plugin_manager
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/plugins/__init__.py", line 28, in <module>
    from .hookspec import hookimpl  # noqa: F401
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/plugins/hookspec.py", line 16, in <module>
    from .types import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/plugins/types.py", line 15, in <module>
    from ..core.solve import Solver
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/core/solve.py", line 30, in <module>
    from ..exceptions import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/exceptions.py", line 15, in <module>
    import requests
  File "/home/positron/anaconda3/lib/python3.8/site-packages/requests/__init__.py", line 45, in <module>
    from .exceptions import RequestsDependencyWarning
  File "/home/positron/anaconda3/lib/python3.8/site-packages/requests/exceptions.py", line 7, in <module>
    from urllib3.exceptions import HTTPError as BaseHTTPError
ModuleNotFoundError: No module named 'urllib3.exceptions'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/positron/anaconda3/bin/conda", line 13, in <module>
    sys.exit(main())
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/cli/main.py", line 109, in main
    return conda_exception_handler(main, *args, **kwargs)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/exception_handler.py", line 389, in conda_exception_handler
    return_value = exception_handler(func, *args, **kwargs)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/exception_handler.py", line 20, in __call__
    return self.handle_exception(exc_val, exc_tb)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/exception_handler.py", line 52, in handle_exception
    from .exceptions import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/conda/exceptions.py", line 15, in <module>
    import requests
  File "/home/positron/anaconda3/lib/python3.8/site-packages/requests/__init__.py", line 45, in <module>
    from .exceptions import RequestsDependencyWarning
  File "/home/positron/anaconda3/lib/python3.8/site-packages/requests/exceptions.py", line 7, in <module>
    from urllib3.exceptions import HTTPError as BaseHTTPError
ModuleNotFoundError: No module named 'urllib3.exceptions'
```
</p>
</details>
even though urllib3 is installed systemwide

```
~$ apt list --installed | grep urllib3
python3-urllib3/jammy,jammy,now 1.26.5-1~exp1
```

<details><summary>When I try to use anaconda-navigator:</summary>
<p>

```
Traceback (most recent call last):
  File "/home/positron/anaconda3/bin/anaconda-navigator", line 7, in <module>
    from anaconda_navigator.app.main import main
  File "/home/positron/anaconda3/lib/python3.8/site-packages/anaconda_navigator/app/main.py", line 19, in <module>
    from anaconda_navigator.app.start import start_app
  File "/home/positron/anaconda3/lib/python3.8/site-packages/anaconda_navigator/app/start.py", line 27, in <module>
    from anaconda_navigator.utils.qthelpers import qapplication
  File "/home/positron/anaconda3/lib/python3.8/site-packages/anaconda_navigator/utils/qthelpers.py", line 18, in <module>
    from anaconda_navigator.utils.linux_scaling import setup_scale_factor_for_linux
  File "/home/positron/anaconda3/lib/python3.8/site-packages/anaconda_navigator/utils/linux_scaling.py", line 8, in <module>
    from anaconda_navigator.utils.conda.launch import run_process
  File "/home/positron/anaconda3/lib/python3.8/site-packages/anaconda_navigator/utils/conda/__init__.py", line 5, in <module>
    from .core import *
  File "/home/positron/anaconda3/lib/python3.8/site-packages/anaconda_navigator/utils/conda/core.py", line 20, in <module>
    import yaml
ModuleNotFoundError: No module named 'yaml'
```
</p>
</details>

<details><summary>When I try to use Spyder:</summary>
<p>

```
Traceback (most recent call last):
  File "/home/positron/anaconda3/bin/spyder", line 11, in <module>
    sys.exit(main())
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/app/start.py", line 256, in main
    mainwindow.main(options, args)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/app/mainwindow.py", line 1823, in main
    mainwindow = create_window(MainWindow, app, splash, options, args)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/app/utils.py", line 289, in create_window
    main.setup()
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/app/mainwindow.py", line 774, in setup
    PLUGIN_REGISTRY.register_plugin(self, PluginClass,
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/api/plugin_registration/registry.py", line 342, in register_plugin
    instance = self._instantiate_spyder5_plugin(
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/api/plugin_registration/registry.py", line 184, in _instantiate_spyder5_plugin
    plugin_instance = PluginClass(main_window, configuration=CONF)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/plugins/completion/plugin.py", line 251, in __init__
    raise e
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/plugins/completion/plugin.py", line 246, in __init__
    Provider = entry_point.resolve()
  File "/home/positron/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py", line 2524, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/plugins/completion/providers/snippets/provider.py", line 21, in <module>
    from spyder.plugins.completion.providers.snippets.conftabs import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/plugins/completion/providers/snippets/conftabs.py", line 25, in <module>
    from spyder.plugins.completion.providers.snippets.widgets import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/plugins/completion/providers/snippets/widgets/__init__.py", line 9, in <module>
    from .snippetsconfig import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/spyder/plugins/completion/providers/snippets/widgets/snippetsconfig.py", line 16, in <module>
    from jsonschema.exceptions import ValidationError
  File "/home/positron/anaconda3/lib/python3.8/site-packages/jsonschema/__init__.py", line 13, in <module>
    from jsonschema._format import FormatChecker
  File "/home/positron/anaconda3/lib/python3.8/site-packages/jsonschema/_format.py", line 11, in <module>
    from jsonschema.exceptions import FormatError
  File "/home/positron/anaconda3/lib/python3.8/site-packages/jsonschema/exceptions.py", line 14, in <module>
    from jsonschema import _utils
  File "/home/positron/anaconda3/lib/python3.8/site-packages/jsonschema/_utils.py", line 12, in <module>
    import importlib_resources as resources  # type: ignore
  File "/home/positron/anaconda3/lib/python3.8/site-packages/importlib_resources/__init__.py", line 3, in <module>
    from ._common import (
  File "/home/positron/anaconda3/lib/python3.8/site-packages/importlib_resources/_common.py", line 10, in <module>
    from .abc import ResourceReader, Traversable
  File "/home/positron/anaconda3/lib/python3.8/site-packages/importlib_resources/abc.py", line 4, in <module>
    from ._compat import runtime_checkable, Protocol
  File "/home/positron/anaconda3/lib/python3.8/site-packages/importlib_resources/_compat.py", line 11, in <module>
    from zipp import Path as ZipPath  # type: ignore
ModuleNotFoundError: No module named 'zipp'
```
</p>
</details>

### Conda Info

```shell
cannot use conda command
```

### Conda Config

```shell
cannot use conda command
```

### Conda list

```shell
cannot use conda command
```
Tried [this](https://docs.anaconda.com/free/anaconda/reference/troubleshooting/#recovering-your-anaconda-installation), didn't fix anything.

### Additional Context

- Operating System / kernel version: Ubuntu 22.04.3 LTS x86_64 / Linux 5.15.0-84-generic 
- conda version: 23.7.3
- Python version: 3.8.18 64-bit

I can still provide files such as `history`, `environments.txt` or `.condarc` if need be.

**Comment by travishathaway** [comment]
Hi @PleaseConverge,

Thanks for filing this issue. At this point the installation is broken and you need to do a reinstall to get it working (although it appears you have already tried that 😅).  Without more detailed information, it is very difficult to determined exactly what happened.

My advice for you would be to try this reinstall again and make sure that you have deactivated and then reactivated the environment again (basically just try this from a new terminal window).

In the future, I would also advise to use the following commands to update your Anaconda Distribution install:

```
# Updates conda itself
conda update conda
# Updates all packages in the distribution
conda update anaconda
```

I'm sorry I can't give a better answer on this.

**Comment by PleaseConverge** [comment]
Thanks for the answer @travishathaway 

Here's what I tried:

- `mv anaconda3 anaconda_old`
- "Deactivate" conda by commenting the `>>> conda initialize >>>` bloc in ~/.zshrc and opening a new terminal
- `zsh ~/Downloads/Anaconda3-2023.07-2-Linux-x86_64.sh`
- Uncomment the `>>> conda initialize >>>` bloc, then `source ~/.zshrc`
- I now have a working fresh install (`conda` command working, `(base)` in the PS1)
- `rsync -a anaconda_old/ anaconda3/`
- `conda` is now broken again, same error as in my first post

I first speculated what happened was that conda being interrupted during the first update resulted in corrupted packages and/or packages that did managed to get fully downloaded were purged from cache/temp (or whatever it is that conda does with packages waiting for install). The second update did not re-download the packages nor check for integrity, and "installed" corrupted packages or just nothing, resulting in package loss.

The problem with this theory is that, checking the history file, none of the packages mentioned in the errors I posted (namely urllib3, yaml and zipp) were part of the last update. So I'm just completely at loss as to how the entire ordeal managed to destroy perfectly good packages that were not involved at all. 

Anyway, I can't even repair my installation so I'm not sure what do at this point. I really don't want to reinstall everything because my environment is several years old at this point, consists of ~50GB of packages, some of which have been been configured in ways I have long forgotten (API wrapper that requires login/key for instance).

Nevertheless, if it does get to this point, all the json are still present in the conda-meta folder. Is there a way to parse those files to streamline the reinstall process and make sure I get back to the exact same environment ?

**Comment by travishathaway** [comment]
Hi @PleaseConverge,

If holding on to that initial base environment is the most important to you, then I would recommend managing it as if it were an environment located at a prefix.

This would require that you install an adjacent conda (e.g. via [Miniconda](https://docs.conda.io/projects/miniconda/en/latest/)). In your debug steps, you would stop at the point before you copied the old installation back to the current one with `rsync -a anaconda_old anaconda`. Going forward you would just have two conda installs: one that's broken and one that works.

Additionally, you could also consider installing something like micromamba which is a conda compatible package manager. This would also allow you to still manage that environment.

I do need to mention that we highly discourage conda users from putting everything in a single environment and instead recommend to organize their environments via named environments. This makes them easier to manage if/when their conda install becomes corrupted.

Hope this helps.

**Comment by PleaseConverge** [comment]
I managed to repair the environment (the original one, `anaconda_old`):

- Using `pip`, I uninstalled/reinstalled the packages that were preventing `conda` from running (urllib3, zipp, tqdm, PyYAML, idna). For some of those, I had to manually delete the corresponding folder in `anaconda3/lib/python3.8/site-packages`, left in a broken state by the interrupted update.
- That gave me a `conda` functional enough to perform `conda install --revision n-1` (revision `n` being the failed update)
- The rollback went through, leaving only a dozen of inconstancies.
- I was still getting an error when trying to `update conda`/`update anaconda`/install a package:
```
Preparing transaction: done
Verifying transaction: failed

RemoveError: 'tqdm' is a dependency of conda and cannot be removed from
conda's operating environment.
```
because tqdm was now installed through pip and it was getting stuck trying to replace it with the conda version I guess ?
- `conda update --force conda` fixed it
- then `conda update anaconda` worked.

Back to square one.

I know that this is all janky, very much not recommanded and there may be a whole bunch of broken stuff left that I just haven't run into yet.

So @travishathaway thank you very much for the tips, chances are I'll have to use them given my dodgy repair. 

>I do need to mention that we highly discourage conda users from putting everything in a single environment and instead recommend to organize their environments via named environments. This makes them easier to manage if/when their conda install becomes corrupted.

Yep while I'm well aware of that now, it wasn't the case when I first started using Anaconda (because I just wanted to use Spyder back then and didn't really care much about the distrib side of things). By the time I learned of those best practices, it was already kind of too late.

Anyway, I'm a bit surprised that `conda` handles insufficient disk space so badly, compared to `apt` for example.
